### PR TITLE
[#5734] feat (gvfs-fuse):  Gvfs-fuse basic FUSE-level implementation and code structure layout

### PR DESCRIPTION
**Pull request by diqiu50**
### What changes were proposed in this pull request?

1. Implement basic FUSE interfaces.
2. Implement filesystem trait and relation structures.

### Why are the changes needed?

Fix: #5734

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

No

### CODE REVIEW

**Comment by FANNG1** [comment]
How do you pass user information to Gravitino?

File: clients/filesystem-fuse/src/filesystem.rs [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+use bytes::Bytes;
+use fuse3::{Errno, FileType, Timestamp};
+
+pub(crate) type Result<T> = std::result::Result<T, Errno>;
+
+/// RawFileSystem interface for the file system implementation. it use by FuseApiHandle
+/// the `file_id` and `parent_file_id` it is the unique identifier for the file system, it is used to identify the file or directory
+/// the `fh` it is the file handle, it is used to identify the opened file, it is used to read or write the file content
+#[async_trait]
+pub(crate) trait RawFileSystem: Send + Sync {
```
[the remark this answers is not in the document]
Comment (diqiu50) [reply clients/filesystem-fuse/src/filesystem.rs:29]: I think the name is OkK, Could you have a good suggestion.

File: clients/filesystem-fuse/src/fuse_api_handle.rs [hunk 2]
```diff
@@ -0,0 +1,445 @@
...
+
+impl<T: RawFileSystem> Filesystem for FuseApiHandle<T> {
+    async fn init(&self, _req: Request) -> fuse3::Result<ReplyInit> {
+        self.local_fs.init().await;
+        Ok(ReplyInit {
+            max_write: NonZeroU32::new(Self::DEFAULT_MAX_WRITE).unwrap(),
+        })
+    }
+
+    async fn destroy(&self, _req: Request) {}
```
[the remark this answers is not in the document]
Comment (diqiu50) [reply clients/filesystem-fuse/src/fuse_api_handle.rs:93]: That's not implemented yet.

File: clients/filesystem-fuse/src/fuse_api_handle.rs [hunk 3]
```diff
@@ -0,0 +1,445 @@
...
+                handle_id.file_id,
+                Directory,
+                Timestamp::from(SystemTime::now()),
+                &self.fs_context,
+            ),
+            generation: 0,
+        })
+    }
+
+    async fn unlink(&self, _req: Request, parent: Inode, name: &OsStr) -> fuse3::Result<()> {
```
[the remark this answers is not in the document]
Comment (diqiu50) [reply clients/filesystem-fuse/src/fuse_api_handle.rs:169]: `unlink` is use by  files. not by the directories

File: clients/filesystem-fuse/src/filesystem.rs [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+    async fn init(&self);
+
+    async fn get_file_path(&self, file_id: u64) -> String;
```
[the remark this answers is not in the document]
Comment (diqiu50) [reply clients/filesystem-fuse/src/filesystem.rs:32]: The `inode id` is special for `inode`,  the `file_id` is generally, it is can look like the  `inode id`

File: clients/filesystem-fuse/src/filesystem.rs [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+
+    async fn read(&self, file_id: u64, fh: u64, offset: u64, size: u32) -> Result<Bytes>;
+
+    async fn write(&self, file_id: u64, fh: u64, offset: u64, data: &[u8]) -> Result<u32>;
+}
+
+/// PathFileSystem is the interface for the file system implementation, it use to interact with other file system
+/// it is used file name or path to operate the file system
+#[async_trait]
+pub(crate) trait PathFileSystem: Send + Sync {
```
[the remark this answers is not in the document]
Comment (diqiu50) [reply clients/filesystem-fuse/src/filesystem.rs:66]: I will commit the usage in the next pr

File: clients/filesystem-fuse/src/fuse_api_handle.rs [hunk 6]
```diff
@@ -0,0 +1,445 @@
...
+    }
+
+    async fn get_modified_file_stat(
+        &self,
+        inode: u64,
+        size: Option<u64>,
+        atime: Option<Timestamp>,
+        mtime: Option<Timestamp>,
+    ) -> Result<FileStat, Errno> {
+        let file_stat = self.local_fs.stat(inode).await?;
```
[the remark this answers is not in the document]
Comment (diqiu50) [reply clients/filesystem-fuse/src/fuse_api_handle.rs:66]: That's not  need clone. I remove it

**Comment by diqiu50** [comment]
> How do you pass user information to Gravitino?

This pr is not touch `Gravitino`, we need consider later

File: clients/filesystem-fuse/src/filesystem.rs [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+    async fn stat(&self, file_id: u64) -> Result<FileStat>;
+
+    /// Get the file stat by parent file id and file name
+    async fn lookup(&self, parent_file_id: u64, name: &str) -> Result<FileStat>;
+
+    /// Read the directory by file id
+    async fn read_dir(&self, dir_file_id: u64) -> Result<Vec<FileStat>>;
+
+    /// Open the file by file id and flags
+    async fn open_file(&self, file_id: u64, flags: u32) -> Result<FileHandle>;
```
Comment (FANNG1) [inline clients/filesystem-fuse/src/filesystem.rs:53]: why return FileHandle not file id?

File: clients/filesystem-fuse/src/fuse_api_handle.rs [hunk 8]
```diff
@@ -0,0 +1,448 @@
...
+        = BoxStream<'a, fuse3::Result<DirectoryEntry>>
+    where
+        T: 'a;
+
+    #[allow(clippy::needless_lifetimes)]
+    async fn readdir<'a>(
+        &'a self,
+        _req: Request,
+        parent: Inode,
+        _fh: u64,
```
Comment (FANNG1) [inline clients/filesystem-fuse/src/fuse_api_handle.rs:260]: why not process `_fh` here?

File: clients/filesystem-fuse/src/fuse_api_handle.rs [hunk 9]
```diff
@@ -0,0 +1,448 @@
...
+                inode: current.parent_file_id,
+                name: "..".into(),
+                kind: Directory,
+                offset: 2,
+            }),
+        ]);
+
+        let combined_stream = relative_paths.chain(entries_stream);
+        Ok(ReplyDirectory {
+            entries: combined_stream.skip(offset as usize).boxed(),
```
Comment (FANNG1) [inline clients/filesystem-fuse/src/fuse_api_handle.rs:292]: The current implement is a little hack, any better solution?

File: clients/filesystem-fuse/src/fuse_api_handle.rs [hunk 10]
```diff
@@ -0,0 +1,446 @@
...
+        _req: Request,
+        parent: Inode,
+        name: &OsStr,
+    ) -> fuse3::Result<ReplyEntry> {
+        let name = name.to_string_lossy();
+        let file_stat = self.local_fs.lookup(parent, &name).await?;
+        Ok(ReplyEntry {
+            ttl: self.default_ttl,
+            attr: fstat_to_file_attr(&file_stat, &self.fs_context),
+            generation: 0,
```
[the remark this answers is not in the document]
Comment (diqiu50) [reply clients/filesystem-fuse/src/fuse_api_handle.rs:107]: The `generation` is use for the filesystem metadata cache of kernel . when the file attributes changed ,   the `generation ` will be increased. But we don't do that . use  the ttl is enough.

File: clients/filesystem-fuse/src/filesystem.rs [hunk 11]
```diff
@@ -0,0 +1,236 @@
...
+
+    /// Open the directory by file id and flags
+    async fn open_dir(&self, file_id: u64, flags: u32) -> Result<FileHandle>;
+
+    /// Create the file by parent file id and file name and flags
+    async fn create_file(&self, parent_file_id: u64, name: &str, flags: u32) -> Result<FileHandle>;
+
+    /// Create the directory by parent file id and file name
+    async fn create_dir(&self, parent_file_id: u64, name: &str) -> Result<FileHandle>;
```
[the remark this answers is not in the document]
Comment (diqiu50) [reply clients/filesystem-fuse/src/filesystem.rs:62]: create file  = new file + open file. 
we should return the file_id and handle_id to up layer

File: clients/filesystem-fuse/src/filesystem.rs [hunk 12]
```diff
@@ -0,0 +1,236 @@
...
+    async fn remove_dir(&self, parent_file_id: u64, name: &str) -> Result<()>;
+
+    /// Close the file by file id and file handle
+    async fn close_file(&self, file_id: u64, fh: u64) -> Result<()>;
+
+    /// Read the file content by file id, file handle, offset and size
+    async fn read(&self, file_id: u64, fh: u64, offset: u64, size: u32) -> Result<Bytes>;
+
+    /// Write the file content by file id, file handle, offset and data
+    async fn write(&self, file_id: u64, fh: u64, offset: u64, data: &[u8]) -> Result<u32>;
```
[the remark this answers is not in the document]
Comment (diqiu50) [reply clients/filesystem-fuse/src/filesystem.rs:80]: Some time, you want to write n bytes. actually the m bytes can  be written. so we need return the number of written bytes.

File: clients/filesystem-fuse/src/filesystem.rs (hunk 7 above)
Comment (diqiu50) [reply clients/filesystem-fuse/src/filesystem.rs:53]: open file should return the file handle id

File: clients/filesystem-fuse/src/fuse_api_handle.rs (hunk 8 above)
Comment (diqiu50) [reply clients/filesystem-fuse/src/fuse_api_handle.rs:260]: directory operation don't need  `fh`.

File: clients/filesystem-fuse/src/fuse_api_handle.rs (hunk 9 above)
Comment (diqiu50) [reply clients/filesystem-fuse/src/fuse_api_handle.rs:292]: We can improve in the other pr.

File: clients/filesystem-fuse/src/fuse_api_handle.rs [hunk 13]
```diff
@@ -0,0 +1,446 @@
...
+                Timestamp::from(SystemTime::now()),
+                &self.fs_context,
+            ),
+            generation: 0,
+        })
+    }
+
+    async fn unlink(&self, _req: Request, parent: Inode, name: &OsStr) -> fuse3::Result<()> {
+        let name = name.to_string_lossy();
+        self.local_fs.remove_file(parent, &name).await?;
```
[the remark this answers is not in the document]
Comment (diqiu50) [reply clients/filesystem-fuse/src/fuse_api_handle.rs:172]: When we does not support hard link, that's the same.

File: clients/filesystem-fuse/src/fuse_api_handle.rs [hunk 14]
```diff
@@ -0,0 +1,446 @@
...
+        _req: Request,
+        inode: Inode,
+        _fh: Option<u64>,
+        set_attr: SetAttr,
+    ) -> fuse3::Result<ReplyAttr> {
+        let new_file_stat = self
+            .get_modified_file_stat(inode, set_attr.size, set_attr.atime, set_attr.mtime)
+            .await?;
+        let attr = fstat_to_file_attr(&new_file_stat, &self.fs_context);
+        self.local_fs.set_attr(inode, &new_file_stat).await?;
```
[the remark this answers is not in the document]
Comment (diqiu50) [reply clients/filesystem-fuse/src/fuse_api_handle.rs:141]: If set_attr failed. error should return

File: clients/filesystem-fuse/src/fuse_api_handle.rs [hunk 15]
```diff
@@ -0,0 +1,446 @@
...
+        })
+    }
+
+    async fn setattr(
+        &self,
+        _req: Request,
+        inode: Inode,
+        _fh: Option<u64>,
+        set_attr: SetAttr,
+    ) -> fuse3::Result<ReplyAttr> {
```
[the remark this answers is not in the document]
Comment (diqiu50) [reply clients/filesystem-fuse/src/fuse_api_handle.rs:136]: Added

File: clients/filesystem-fuse/src/filesystem.rs [hunk 16]
```diff
@@ -0,0 +1,236 @@
...
+    /// Create the file by parent file path and file name and flags
+    async fn create_file(
+        &self,
+        parent: &str,
+        name: &str,
+        flags: OpenFileFlags,
+    ) -> Result<OpenedFile>;
+
+    /// Create the directory by parent file path and file name
+    async fn create_dir(&self, parent: &str, name: &str) -> Result<OpenedFile>;
```
Comment (FANNG1) [inline clients/filesystem-fuse/src/filesystem.rs]: should return open file here?

File: clients/filesystem-fuse/src/filesystem.rs [hunk 17]
```diff
@@ -0,0 +1,236 @@
...
+            uid,
+            gid,
+            default_file_perm: 0o644,
+            default_dir_perm: 0o755,
+            block_size: 4 * 1024,
+        }
+    }
+}
+
+// FileStat is the file stat for the file system.
```
Comment (FANNG1) [inline clients/filesystem-fuse/src/filesystem.rs:156]: There are too many file systems, please specify which file system.

File: clients/filesystem-fuse/src/filesystem.rs [hunk 18]
```diff
@@ -0,0 +1,236 @@
...
+    pub(crate) atime: Timestamp,
+
+    // file modify time
+    pub(crate) mtime: Timestamp,
+
+    // file create time
+    pub(crate) ctime: Timestamp,
+
+    // file link count
+    pub(crate) nlink: u32,
```
Comment (FANNG1) [inline clients/filesystem-fuse/src/filesystem.rs:190]: do we need `nlink`?

File: clients/filesystem-fuse/src/fuse_api_handle.rs [hunk 19]
```diff
@@ -0,0 +1,507 @@
...
+    }
+
+    async fn setattr(
+        &self,
+        _req: Request,
+        inode: Inode,
+        fh: Option<u64>,
+        set_attr: SetAttr,
+    ) -> fuse3::Result<ReplyAttr> {
+        // check the opened file file_id is the same as the inode
```
Comment (FANNG1) [inline clients/filesystem-fuse/src/fuse_api_handle.rs:137]: please add the comment to the interface of `valid_file_id `, no need to add the comment for the invoked places.
change the comment to `Verify the equality between the file_id of the opened file and the inode.`？

**Comment by FANNG1** [comment]
LGTM, except minor comments

File: clients/filesystem-fuse/src/filesystem.rs (hunk 18 above)
Comment (diqiu50) [reply clients/filesystem-fuse/src/filesystem.rs:190]: That's a reference count of the file. we may use it later.

File: clients/filesystem-fuse/src/filesystem.rs (hunk 17 above)
Comment (diqiu50) [reply clients/filesystem-fuse/src/filesystem.rs:156]: Update

File: clients/filesystem-fuse/src/filesystem.rs [hunk 20]
```diff
@@ -32,33 +32,33 @@ pub(crate) type Result<T> = std::result::Result<T, Errno>;
...
-    /// Open the directory by file id and flags
+    /// Open the directory by file id and flags, if successful, return the file handle
     async fn open_dir(&self, file_id: u64, flags: u32) -> Result<FileHandle>;
 
-    /// Create the file by parent file id and file name and flags
+    /// Create the file by parent file id and file name and flags, if successful, return the file handle
     async fn create_file(&self, parent_file_id: u64, name: &str, flags: u32) -> Result<FileHandle>;
 
-    /// Create the directory by parent file id and file name
+    /// Create the directory by parent file id and file name, if successful, return the file id
```
Comment (FANNG1) [inline clients/filesystem-fuse/src/filesystem.rs:61]: How about defining a new type `FileID` not using u64 directly here?

File: clients/filesystem-fuse/src/filesystem.rs [hunk 21]
```diff
@@ -0,0 +1,241 @@
...
+/// the `file id` and `handle_id` need to mapping the `ino`/`inode` and `fh` in the fuse3
+#[async_trait]
+pub(crate) trait RawFileSystem: Send + Sync {
+    /// Init the file system
+    async fn init(&self) -> Result<()>;
+
+    /// Get the file path by file id, if the file id is valid, return the file path
+    async fn get_file_path(&self, file_id: u64) -> String;
+
+    /// Validate the file id and file handle, if the file id and file handle is valid, return success
```
Comment (FANNG1) [inline clients/filesystem-fuse/src/filesystem.rs:40]: please add more comments here about how to valid

File: clients/filesystem-fuse/src/filesystem.rs (hunk 18 above)
Comment (FANNG1) [reply clients/filesystem-fuse/src/filesystem.rs:190]: Could you provide an example about how to use it ?

File: clients/filesystem-fuse/src/filesystem.rs (hunk 20 above)
Comment (diqiu50) [reply clients/filesystem-fuse/src/filesystem.rs:61]: use u64 is ok.  It shouldn't be necessary.

File: clients/filesystem-fuse/src/filesystem.rs (hunk 18 above)
Comment (diqiu50) [reply clients/filesystem-fuse/src/filesystem.rs:190]: This can be considered in a later PR. Let's first push the framework so others can join in.

**Review by FANNG1 — APPROVED** [review]

**Pull request merged**